File: src/leetcode/python/sliding_window/max_profit.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # initialize pointers
        l, r = 0, 1
        max_profit = 0

        while r < len(prices):
            current_profit = prices[r] - prices[l]
            if prices[l] < prices[r]:
                max_profit = max(max_profit, current_profit)
            else:
                l = r  # if we find a better point to buy
            r += 1

        return max_profit


if __name__ == "__main__":
    sol = Solution()
    print(sol.maxProfit([7, 1, 5, 3, 6, 4]))  # output  5
    print(sol.maxProfit([7, 6, 4, 3, 1]))  # output 0

File: src/leetcode/julia/sliding_window/MaxProfit.jl
function MaxProfit(prices::Vector{Int})::Int
    # trackers
    left = 1
    right = 2
    max_profit = 0 

    while right <= length(prices)
        current_profit = prices[right] - prices[left]
        if prices[right] > prices[left]
            max_profit = max(max_profit, current_profit)
        else
            left = right
        end
        right += 1
    end
    return max_profit
end